.jenkins/Jenkinsfile

```diff
@@ -22,517 +22,447 @@
+  'cqlsh-test': [type: 'test', splits: false],
+  'fqltool-test': [type: 'test', splits: false],
+  'test-cdc': [type: 'test', splits: splits],
+  'test': [type: 'test', splits: splits],
+  'test-trie': [type: 'test', splits: splits],
+  'test-compression': [type: 'test', splits: splits],
+  'stress-test': [type: 'test', splits: splits],
+  'test-burn': [type: 'test', splits: splits],
+  'long-test': [type: 'test', splits: splits],
+  'test-oa': [type: 'test', splits: splits],
+  'test-system-keyspace-directory': [type: 'test', splits: splits],
+  'jvm-dtest': [type: 'test', splits: splits, dtest: true],
+  'jvm-dtest-upgrade': [type: 'test', splits: splits, dtest: true],
+  'dtest': [type: 'test', splits: splits, dtest: true],
+  'dtest-novnode': [type: 'test', splits: splits, dtest: true],
+  'dtest-offheap': [type: 'test', splits: splits, dtest: true],
+  'dtest-large': [type: 'test', splits: splits, dtest: true],
+  'dtest-large-novnode': [type: 'test', splits: splits, dtest: true],
+  'dtest-upgrade': [type: 'test', splits: splits, dtest: true],
+  'dtest-upgrade-novnode': [type: 'test', splits: splits, dtest: true],
+  'dtest-upgrade-large': [type: 'test', splits: splits, dtest: true],
+  'dtest-upgrade-novnode-large': [type: 'test', splits: splits, dtest: true],
+]
+
+// define matrix axes
+def Map matrix_axes = [
+    arch: archsSupported,
+    jdk: jdksSupported,
+    python: pythonsSupported,
+    cython: ['yes', 'no'],
+    step: stepsMap.keySet(),
+    split: (1..splits).toList()
+]
+
+def List _axes = getMatrixAxes(matrix_axes).findAll { axis ->
+    (isArchEnabled(axis['arch'])) && // skip disabled archs
+    (isJdkEnabled(axis['jdk'])) && // skip disabled jdks
+    (isStageEnabled(axis['step'])) && // skip disabled steps
+    !(axis['python'] != pythonDefault && axis['step'] != 'cqlsh-test') && // Use only python 3.7 for all tests but cqlsh-test
+    !(axis['cython'] != 'no' && axis['step'] != 'cqlsh-test') && // cython only for cqlsh-test, disable for others
+    !(axis['arch'] != 'amd64' && axis['step'] == 'cqlsh-test') && // run cqlsh-test only on amd64
+    !(axis['jdk'] != jdkDefault && axis['step'] == 'cqlsh-test') && // run cqlsh-test only with jdk11
+    // Disable splits for all but proper stages
+    !(axis['split'] != 1 && !stepsMap.findAll { entry -> entry.value.containsKey('splits') && entry.value.splits }.keySet().contains(axis['step'])) &&
+    // run dtests only on amd64
+    !(axis['arch'] != 'amd64' && stepsMap.findAll { entry -> entry.value.containsKey('dtest') && entry.value.dtest }.keySet().contains(axis['step']))
+}
+
+// Prepare tasks
+def Map tasks = [
+  jars: [failFast: false],
+  tests: [failFast: false],
+]
+for (def axis in _axes) {
+  def cell = axis
+  def name = getStepName(cell, stepsMap[cell.step])
+  tasks[cell.step == "jar" ? "jars" : "tests"][name] = { ->
+    "${stepsMap[cell.step].type}"(stepsMap[cell.step], cell) // build() or test()
+  }
+}
+
```

Comment:
how much of the above scripting can we move to below line 180 (alongside the other scripting support code) ?

Reply:
I will tackle this later. I have some ideas…

Reply:
Well, nearly all of it. Should I?

Reply:
I'd keep stepsMap and matrix definition upstairs for sure, maybe excludes too... Just for visibility. But no pushing, can move it down.